packages/store/ts/config/v2/tableFull.ts

```diff
@@ -0,0 +1,137 @@
+import { conform, evaluate } from "@arktype/util";
+import { get, hasOwnKey } from "./generics";
+import { SchemaInput, isSchemaInput, resolveSchema } from "./schema";
+import { AbiTypeScope, getStaticAbiTypeKeys } from "./scope";
+import { isStaticAbiType } from "@latticexyz/schema-type";
+
+export type TableFullInput<
+  schema extends SchemaInput<scope> = SchemaInput,
+  scope extends AbiTypeScope = AbiTypeScope,
+  primaryKey extends ValidKeys<schema, scope> = ValidKeys<schema, scope>,
+> = {
+  schema: schema;
+  primaryKey: primaryKey;
+};
+
+export type ValidKeys<schema extends SchemaInput<scope>, scope extends AbiTypeScope> = readonly [
+  getStaticAbiTypeKeys<schema, scope>,
+  ...getStaticAbiTypeKeys<schema, scope>[],
+];
+
+function getValidKeys<schema extends SchemaInput<scope>, scope extends AbiTypeScope>(
+  schema: schema,
+  scope: scope = AbiTypeScope as scope,
+): ValidKeys<schema, scope> {
+  return Object.entries(schema)
+    .filter(([, internalType]) => hasOwnKey(scope.types, internalType) && isStaticAbiType(scope.types[internalType]))
+    .map(([key]) => key) as unknown as ValidKeys<schema, scope>;
+}
+
+export function isValidPrimaryKey<schema extends SchemaInput<scope>, scope extends AbiTypeScope>(
+  primaryKey: unknown,
+  schema: schema,
+  scope: scope = AbiTypeScope as scope,
+): primaryKey is ValidKeys<schema, scope> {
+  return (
+    Array.isArray(primaryKey) &&
+    primaryKey.every(
+      (key) =>
+        hasOwnKey(schema, key) && hasOwnKey(scope.types, schema[key]) && isStaticAbiType(scope.types[schema[key]]),
+    )
+  );
+}
+
+export function isTableFullInput(input: unknown): input is TableFullInput {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    hasOwnKey(input, "schema") &&
+    hasOwnKey(input, "primaryKey") &&
+    Array.isArray(input["primaryKey"])
+  );
+}
+
+export type validateKeys<validKeys extends PropertyKey, keys> = {
+  [i in keyof keys]: keys[i] extends validKeys ? keys[i] : validKeys;
+};
+
+export type validateTableFull<input, scope extends AbiTypeScope = AbiTypeScope> = {
+  [key in keyof input]: key extends "primaryKey"
+    ? validateKeys<getStaticAbiTypeKeys<conform<get<input, "schema">, SchemaInput<scope>>, scope>, input[key]>
+    : key extends "schema"
+      ? conform<input[key], SchemaInput<scope>>
+      : input[key];
+};
+
+export function validateTableFull<input, scope extends AbiTypeScope = AbiTypeScope>(
+  input: input,
+  scope: scope = AbiTypeScope as scope,
+): void {
```

Comment:
would it be useful to have this narrow type type as well?
```suggestion
): asserts input is TableFullInput {
```

Reply:
updated